teps/0101-env-in-pod-template.md

```diff
@@ -0,0 +1,407 @@
+   variables at the `PipelineRun` and `TaskRun` can reuse the same task with different Cloud Provider credentials.
+   Kubernetes API `valueFrom` can also refe to values in the pod labels/annotations for specific Kubernetes cluster
+   information such as namespace, application labels, and service annotations.
+
+## Requirements
+
+<!--
+Describe constraints on the solution that must be met. Examples might include
+performance characteristics that must be met, specific edge cases that must
+be handled, or user scenarios that will be affected and must be accomodated.
+-->
+
+1. In the first case, common environment variables can be defined in a single place on a `PipelineRun` level.
+   Values can be specified as literals or through references.
+   Variables defined on a `PipelineRun` or `TaskRun` level are then available in all steps.
+   That allows to significantly reduce the size of `PipelineRun` and `TaskRun` resource,
+   by excluding the common environment variables like: static global settings, common values coming from metadata, ...
+
+2. Secondly, environment variables defined in steps can be easily overwritten by the ones from `PipelineRun` and `TaskRun`.
+  With that, common settings like API keys, connection details, ... can be optionally overwritten in a single place. For
+  example, if both `PipelineRun` and task `StepTemplate` has environment variables PROXY, it will overwrite the task
+  `StepTemplate` with the environment variable values inside `PipelineRun`.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Common environment variables can be defined in a single place on a `PipelineRun` level.
+Values can be specified as literals or through references, e.g.:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: mypipelinerun
+spec:
+  podTemplate:
+    env:
+    - name: TOKEN_PATH
+      value: /opt/user-token
+    - name: TKN_PIPELINE_RUN
+      valueFrom:
+        fieldRef:
+          fieldPath: metadata.labels['tekton.dev/pipelineRun']
+```
+
+Environment variables defined in steps can be easily overwritten by the ones from a `TaskRun`, e.g.:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: mytask
+  namespace: default
+spec:
+  steps:
+    - name: echo-msg
+      image: ubuntu
+      command: ["bash", "-c"]
+      args: ["echo $MSG"]
+      envs:
+      - name: "MSG"
+        value: "Default message"
+---
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: mytaskrun
+  namespace: default
+spec:
+  taskRef:
+    name: mytask
+  podTemplate:
+    envs:
+      - name: "MSG"
+        value: "Overwritten message"
+```
+
```

Comment:
thank you for the example above! it demonstrates meeting requirement 2, can we add an example for requirement 1 as well?

Reply:
sure no problem